### PR TITLE
Adding a snippet for metrics

### PR DESCRIPTION
**Pull request by pabloem**
Adding a snippet to use in metrics documentation.

### CODE REVIEW

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 1]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+      | beam.io.ReadFromText(
+          'gs://dataflow-samples/shakespeare/kinglear.txt')
+      | 'ExtractWords' >> beam.FlatMap(lambda x: re.findall(r'[A-Za-z\']+', x))
+      | beam.combiners.Count.PerElement()
+      | 'FilterText' >> beam.ParDo(FilterTextFn('Flourish|stomach')))
+  # [END metrics_usage_example]
+
+  output = (filtered_words
+            | 'format' >> beam.Map(lambda (word, c): '%s: %s' % (word, c))
+            | 'Write' >> beam.io.WriteToText('gs://my-bucket/counts.txt'))
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Don't write to files unless you must. (In fact, this whole output computation could be moved.)

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 2]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+
+  p.visit(SnippetUtils.RenameFiles(renames))
+  p.run()
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Verify the results?

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 3]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+      word, _ = element
+      if re.match(self.pattern, word):
+        # Add 1 to the custom metric counter matched_words
+        self.matched_words.inc()
+        yield element
+      else:
+        # Add 1 to the custom metric counter umatched_words
+        self.umatched_words.inc()
+
+  p = TestPipeline()  # Use TestPipeline for testing.
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Generally we don't put the (common) pipeline construction code into the snippet.

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 4]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+  filtered_words = (
+      p
+      | beam.io.ReadFromText(
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Don't read from a file unless you're trying to demonstrate file reads. Just start with Create(...) (and outside of the example).

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 5]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+          'gs://dataflow-samples/shakespeare/kinglear.txt')
+      | 'ExtractWords' >> beam.FlatMap(lambda x: re.findall(r'[A-Za-z\']+', x))
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: ExtractWords and CountPerElement seem unrelated to what you're trying to show here (and complicate your process method as well). Instead I'd just write a DoFn that consumes a PCollection of words.

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 6]
```diff
@@ -1170,3 +1170,47 @@ def expand(self, pcoll):
...
+      # A custom metric can track values in your pipeline as it runs. Create
+      # custom metrics matched_words and unmatched_words.
+      self.matched_words = Metrics.counter(self.__class__, 'matched_words')
+      self.umatched_words = Metrics.counter(self.__class__, 'umatched_words')
+
+    def process(self, element):
+      word, _ = element
+      if re.match(self.pattern, word):
+        # Add 1 to the custom metric counter matched_words
+        self.matched_words.inc()
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: It might be more interesting to come up with an example that doesn't have mutually exclusive counters, to illustrate one can increment whichever counters in whatever order you want. Maybe a distribution of word lengths would be interesting?

**Comment by pabloem** [comment]
Thanks for the comments @robertwb. I've added a distribution and made both metrics non-mutually exclusive. Also tried to format the snippet to follow conventions set in the file.

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 7]
```diff
@@ -1171,46 +1171,54 @@ def expand(self, pcoll):
...
+        self.pair_counter.inc()
         yield element
-      else:
-        # Add 1 to the custom metric counter umatched_words
-        self.umatched_words.inc()
 
-  p = TestPipeline()  # Use TestPipeline for testing.
-  filtered_words = (
+
+  filtered_numbers = (
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: Move this up before the [START metrics_usage_example], just have the definition of the DoFn and `words | beam.ParDo(...)` here.

File: sdks/python/apache_beam/examples/snippets/snippets.py [hunk 8]
```diff
@@ -1171,46 +1171,54 @@ def expand(self, pcoll):
         | beam.CombinePerKey(sum))
 # [END model_library_transforms_count]
 
-def examples_troubleshooting_metrics(renames):
+def examples_troubleshooting_metrics(number_list, output_path):
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets.py]: You can add a test to snippets_test, or just put this whole thing in a test method there.

**Comment by pabloem** [comment]
Thanks again for the feedback Robert. I moved it to test file, added checks, and removed the extra stuff from the tagged snippets themselves.

File: sdks/python/apache_beam/examples/snippets/snippets_test.py [hunk 9]
```diff
@@ -693,6 +695,64 @@ def test_model_co_group_by_key_tuple(self):
...
+        self.word_len_dist = Metrics.distribution(self.__class__,
+                                                  'word_len_dist')
+        self.matched_words = Metrics.counter(self.__class__, 'matched_words')
+
+      def process(self, element):
+        word = element
+        self.word_len_dist.update(len(word))
+        if re.match(self.pattern, word):
+          self.matched_words.inc()
+          yield element
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets_test.py]: Nit: this is equal to the size of filtered_words; perhaps let the counter be omitted/dropped words?

File: sdks/python/apache_beam/examples/snippets/snippets_test.py [hunk 10]
```diff
@@ -693,6 +695,64 @@ def test_model_co_group_by_key_tuple(self):
...
+
+    filtered_words = (
+        words | 'FilterText' >> beam.ParDo(FilterTextFn('Flourish|stomach')))
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets_test.py]: People might scratch their heads as to what this means. I might do something like "s.*" or ".*boo.*" (and change the wording below).

**Comment by pabloem** [comment]
Good points. I've updated the example.

File: sdks/python/apache_beam/examples/snippets/snippets_test.py [hunk 11]
```diff
@@ -693,6 +695,67 @@ def test_model_co_group_by_key_tuple(self):
...
+      logging.info('There were %d words that did not match the filter.',
+                   custom_counter[0].committed)
+    # [END metrics_check_values_example]
+
+    # There should be 4 words that did not match
+    self.assertEqual(custom_counter[0].committed, 4)
+    # The shortest word is 3 characters, the longest is 6
+    self.assertEqual(custom_distribution[0].committed.min, 3)
+    self.assertEqual(custom_distribution[0].committed.max, 6)
+
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/snippets/snippets_test.py]: apache_beam/examples/snippets/snippets_test.py:755:3: E303 too many blank lines (2)
